p2p/conn_tracker.go

```diff
@@ -0,0 +1,75 @@
+package p2p
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+)
+
+type connectionTracker interface {
+	AddConn(net.IP) error
+	CheckConn(net.IP) (int, bool)
+	RemoveConn(net.IP)
+}
+
+type connTrackerImpl struct {
+	cache       map[string]uint
+	lastConnect map[string]time.Time
+	mutex       sync.RWMutex
+	max         uint
+	window      time.Duration
+}
+
+func newConnTracker(max uint, window time.Duration) connectionTracker {
+	return &connTrackerImpl{
+		cache:       make(map[string]uint),
+		lastConnect: make(map[string]time.Time),
+		max:         max,
+	}
+}
+
+func (rat *connTrackerImpl) AddConn(addr net.IP) error {
```

Comment:
rat? haha

Reply:
oh, lol. I definitely named the type "requestAccessTracker" at first, but then changed it.